### PR TITLE
Premint

### PR DESCRIPTION
**Pull request by oveddan**
* Premint - delegated contract & token creation, & minting
* First minter rewards
* Determinstic contract addresses

### CODE REVIEW

**Comment by oveddan** [comment]
Current dependencies on/for this PR:
* main
  * **PR #161** <a href="https://app.graphite.dev/github/pr/ourzora/zora-1155-contracts/161" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/ourzora/zora-1155-contracts/161?utm_source=stack-comment).

File: src/premint/README.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
+# Preminter
```
Comment (oveddan) [inline src/premint/README.md:1]: this should be made up to date before merging to master

File: src/premint/ZoraCreator1155Attribution.sol [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+                )
+            );
+    }
+
+    function _stringHash(string calldata value) private pure returns (bytes32) {
+        return keccak256(bytes(value));
+    }
+
+    // todo: move to its own contract
+    error MintNotYetStarted();
```
Comment (oveddan) [inline src/premint/ZoraCreator1155Attribution.sol:161]: lets do this todo, and just reference IZora1155Errors.MintNotYetStarted

File: src/premint/ZoraCreator1155Attribution.sol [hunk 3]
```diff
@@ -0,0 +1,240 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
+
+import {IMinter1155} from "../interfaces/IMinter1155.sol";
+import {IZoraCreator1155} from "../interfaces/IZoraCreator1155.sol";
+import {ICreatorRoyaltiesControl} from "../interfaces/ICreatorRoyaltiesControl.sol";
+import {ECDSAUpgradeable} from "@zoralabs/openzeppelin-contracts-upgradeable/contracts/utils/cryptography/ECDSAUpgradeable.sol";
+import {ZoraCreatorFixedPriceSaleStrategy} from "../minters/fixed-price/ZoraCreatorFixedPriceSaleStrategy.sol";
+
+struct ContractCreationConfig {
```
Comment (oveddan) [inline src/premint/ZoraCreator1155Attribution.sol:10]: this doesn't quite belong in this file, as it is primarily used to create new contracts and get a deterministic address based on these fields

File: src/premint/ZoraCreator1155PremintExecutor.sol [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+import {IHasContractName} from "../interfaces/IContractMetadata.sol";
+import {IZoraCreator1155} from "../interfaces/IZoraCreator1155.sol";
+import {IZoraCreator1155Errors} from "../interfaces/IZoraCreator1155Errors.sol";
+import {IZoraCreator1155Factory} from "../interfaces/IZoraCreator1155Factory.sol";
+import {SharedBaseConstants} from "../shared/SharedBaseConstants.sol";
+import {ZoraCreatorFixedPriceSaleStrategy} from "../minters/fixed-price/ZoraCreatorFixedPriceSaleStrategy.sol";
+import {IMinter1155} from "../interfaces/IMinter1155.sol";
+import {PremintConfig, ContractCreationConfig, TokenCreationConfig, ZoraCreator1155Attribution} from "./ZoraCreator1155Attribution.sol";
+
+/// @title Enables creation of and minting tokens on Zora1155 contracts transactions using eip-712 signatures.
```
Comment (oveddan) [inline src/premint/ZoraCreator1155PremintExecutor.sol:16]: add Enjoy graphic (and inherit from it)

File: src/premint/ZoraCreator1155PremintExecutor.sol [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+        );
+    }
+
+    function _getOrCreateContract(ContractCreationConfig calldata contractConfig) private returns (IZoraCreator1155 tokenContract, bool isNewContract) {
+        address contractAddress = getContractAddress(contractConfig);
+        // first we see if the code is already deployed for the contract
+        isNewContract = contractAddress.code.length == 0;
+
+        if (isNewContract) {
+            // if address doesnt exist for hash, createi t
```
Comment (oveddan) [inline src/premint/ZoraCreator1155PremintExecutor.sol:103]: fix typo

File: src/premint/ZoraCreator1155PremintExecutor.sol [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    }
+
+    // upgrade related functionality
+
+    /// @notice The name of the contract for upgrade purposes
+    function contractName() external pure returns (string memory) {
+        return "ZORA 1155 Premint Executor";
+    }
+
+    // upgrade functionality
```
Comment (oveddan) [inline src/premint/ZoraCreator1155PremintExecutor.sol:180]: duplicate comment

**Review by oveddan — COMMENTED** [review]
some minor fixes should be worked on before merging to master

**Review by iainnash — APPROVED** [review]

File: package/batchPublish.test.ts [hunk 7]
```diff
@@ -281,7 +280,7 @@ describe("ZoraCreator1155Preminter", () => {
         [collectorAccount]
       );
 
-      const zoraMintFee = chainConfigs[zora.id].MINT_FEE_AMOUNT;
+      const zoraMintFee = parseEther("0.0007");
```
Comment (iainnash) [inline package/batchPublish.test.ts:283]: Should this be `0.000777`?

Comment (oveddan) [reply package/batchPublish.test.ts:283]: i delete this test for now

Comment (oveddan) [reply package/batchPublish.test.ts:283]: yes, fixing

**Pull request merged**